### PR TITLE
Add Least Privilege Score Check to Policy MS.AAD.7.2v1

### LINKED ISSUE #760: Implement ScubaGear policy check for MS.AAD.7.2v1
# 💡 Summary #

The policy MS.AAD.7.2v1 "Privileged users SHALL be provisioned with finer-grained roles instead of Global Administrator" is currently a manual check in ScubaGear. The scope of this issue is to implement a ScubaGear policy check based on a new least privilege formula that we [designed and tested in Flipper](https://github.com/cisagov/ScubaGear/issues/548#issuecomment-1852406542).

## Acceptance criteria ##

How do we know when this work is done?

- [x] Changes to AAD provider and Rego are complete and tested by the developer.
- [x] Unit tests have been modified and pass.
- [x] Functional tests have been modified and pass.

### PR DESCRIPTION
**Pull request by ssatyapal123**
#  #
  

## 🗣 Description ##

This pull request modifies the AAD Rego code for policy MS.AAD.7.2 "Privileged users SHALL be provisioned with finer-grained roles instead of Global Administrator." by implementing a custom calculation for computing the tenant's least privilege check.  The least privilege score is calculated as the ratio of the number of users with Global Admin privileges to the number of users with other highly privileged roles. This change allows ScubaGear to automatically check for the least privilege score versus having the user manually check the Secure Score value for the action named **Use least privileged administrative roles**. This PR also includes Rego unit tests to check the logic in the Rego code.

Closes #760 

## 🧪 Testing ##
I tested this against the G5 and E5 tenants which each had different scores. In addition, the unit tests check for the different pass and fail cases. To test this, run the ScubaGear tool against aad and check the report output for policy MS.AAD.7.2. To verify the result is correct, review the privileged_users array in the ProviderSettingsExport.json to count the number of users with Global Admin privileges and the number with other privileged roles but not Global Admin, and compute the ratio. This policy is dependent on MS.AAD.7.1 such that if MS.AAD.7.1 fails, MS.AAD.7.2 will automatically fail. This is because it is not meaningful to compute a least privilege score if the number of Global Admins exceeds the limit specified in the policy for MS.AAD.7.1,

## ✅ Pre-approval checklist ##

- [x] This PR has an informative and human-readable title.
- [x] PR targets the correct parent branch (e.g., main or release-name) for merge.
- [x] Changes are limited to a single goal - *eschew scope creep!*
- [x] Changes are sized such that they do not touch excessive number of files.
- [x] *All* future TODOs are captured in issues, which are referenced in code comments.
- [x] These code changes follow the ScubaGear [content style guide](https://github.com/cisagov/ScubaGear/blob/main/CONTENTSTYLEGUIDE.md).
- [x] Related issues these changes resolve are linked preferably via [closing keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).
- [x] All relevant type-of-change labels added.
- [x] All relevant project fields are set.
- [x] All relevant repo and/or project documentation updated to reflect these changes.
- [x] Unit tests added/updated to cover PowerShell and Rego changes.
- [x] Functional tests added/updated to cover PowerShell and Rego changes.
- [x] All relevant functional tests passed.
- [ ] All automated checks (e.g., linting, static analysis, unit/smoke tests) passed.

## ✅ Pre-merge checklist ##

- [ ] PR passed smoke test check.
- [ ] Feature branch has been rebased against changes from parent branch, as needed

  Use `Rebase branch` button below or use [this](https://www.digitalocean.com/community/tutorials/how-to-rebase-and-update-a-pull-request) reference to rebase from the command line.
- [ ] Resolved all merge conflicts on branch
- [ ] Notified merge coordinator that PR is ready for merge via comment mention

## ✅ Post-merge checklist ##

- [ ] Feature branch deleted after merge to clean up repository.
- [ ] Verified that all checks pass on parent branch (e.g., main or release-name) after merge.

### CODE REVIEW

File: sample-config-files/aad-config.yaml [hunk 1]
```diff
@@ -27,26 +27,26 @@ OutFolderName: M365BaselineConformance
 OutProviderFileName: ProviderSettingsExport
 OutRegoFileName: TestResults
 OutReportName: BaselineReports
-GlobalVars: # For cross product variables
+GlobalVars:  # For cross product variables
```
Comment (Sloane4) [inline sample-config-files/aad-config.yaml:30]:
```suggestion
GlobalVars: # For cross product variables
```

File: sample-config-files/aad-config.yaml [hunk 2]
```diff
@@ -27,26 +27,26 @@ OutFolderName: M365BaselineConformance
...
         - 08adb07a-956f-450e-b41c-81e92e3db2c4
         - 8454f405-3b29-4102-b888-315c4e3de2d0
-  MS.AAD.2.1v1: *CommonExclusions
-  MS.AAD.2.3v1: *CommonExclusions
-  MS.AAD.3.1v1: *CommonExclusions
-  MS.AAD.3.2v1: *CommonExclusions
+  MS.AAD.2.1v1:  *CommonExclusions
+  MS.AAD.2.3v1:  *CommonExclusions
+  MS.AAD.3.1v1:  *CommonExclusions
+  MS.AAD.3.2v1:  *CommonExclusions
```
Comment (Sloane4) [inline sample-config-files/aad-config.yaml:44]:
```suggestion
  MS.AAD.2.1v1: *CommonExclusions
  MS.AAD.2.3v1: *CommonExclusions
  MS.AAD.3.1v1: *CommonExclusions
  MS.AAD.3.2v1: *CommonExclusions
```

File: sample-config-files/aad-config.yaml [hunk 3]
```diff
@@ -17,7 +17,7 @@ Description: |
   - Namespace for each policy item within a product for variables related only to one policy item (i.e., Aad.Policy2_1)
   - Use of YAML anchors and aliases following DRY (Don't Repeat Yourself) principle for repeated values and sections
 ProductNames:
-  - aad
+- aad
```
Comment (Sloane4) [inline sample-config-files/aad-config.yaml:20]: I don't see any changes to this file that affects the code. What is the purpose of these changes?

File: PowerShell/ScubaGear/Rego/AADConfig.rego [hunk 4]
```diff
@@ -666,25 +672,44 @@ tests contains {
...
 #--
 
-# At this time we are unable to test for 7.2v1
+# Save all users that don't have Global Admin role
+NotGlobalAdmins contains User.DisplayName if {
+    some User in input.privileged_users
+    not "Global Administrator" in User.roles
+}
+
+default GetScoreDescription := "No privileged users that are NOT Global Admin; Least Privilege Score = 100%"
```
Comment (gdasher) [inline PowerShell/ScubaGear/Rego/AADConfig.rego:688]: In this case, the least score is really infinite, right?  100% would be a passing score but this is a failure mode (and the condition implemented in the actual policy test will fail, as expected).

Aside: we will have to see, but there may be a case for tiny agencies where |{Global Admins}| = 1 or 2 and there is no need for any other admins.  As designed, this would fail the score.  For now, I think its fine but we may need to fix this later based on additional learnings.

**Review by adhilto — COMMENTED** [review]
Nothing critical, just some comments on code style.

File: PowerShell/ScubaGear/Testing/Unit/Rego/AAD/AADConfig_07_test.rego [hunk 5]
```diff
@@ -123,15 +122,152 @@ test_PrivilegedUsers_Incorrect_V2 if {
...
+                "DisplayName": "Test Name2",
+                "roles": [
+                    "Privileged Role Administrator",
+                    "Global Administrator"
+                ]
+            }
+        }
+    }
+
+    ReportDetailStr := "Requirement not met: No privileged users that are NOT Global Admin; Least Privilege Score is undefined"
```
[the remark this answers is not in the document]
Comment (ssatyapal123) [reply PowerShell/ScubaGear/Testing/Unit/Rego/AAD/AADConfig_07_test.rego]: shortened string

File: PowerShell/ScubaGear/Rego/AADConfig.rego [hunk 6]
```diff
@@ -655,6 +655,12 @@ GlobalAdmins contains User.DisplayName if {
     "Global Administrator" in User.roles
 }
 
+#Set conditions under which this policy will pass
+GlobalAdminConditions := [
```
[the remark this answers is not in the document]
Comment (ssatyapal123) [reply PowerShell/ScubaGear/Rego/AADConfig.rego]: Great suggestions, changes made.

File: PowerShell/ScubaGear/Rego/AADConfig.rego [hunk 7]
```diff
@@ -666,25 +673,44 @@ tests contains {
...
+    not "Global Administrator" in User.roles
+}
+
+default GetScoreDescription := "All privileged users are Global Admin"
+GetScoreDescription := concat("", ["Least Privilege Score = ", format_int(x,10), "%"]) if {
+    count(NotGlobalAdmins) > 0
+    x := count(GlobalAdmins)/count(NotGlobalAdmins)*100
+}
+
+#calculate least privilege score as ratio of priv users with global admin role to priv users without global admin role
```
Comment (adhilto) [inline PowerShell/ScubaGear/Rego/AADConfig.rego:695]: One more instance of this
```suggestion
# calculate least privilege score as ratio of priv users with global admin role to priv users without global admin role
```

File: PowerShell/ScubaGear/Rego/AADConfig.rego [hunk 8]
```diff
@@ -666,25 +673,44 @@ tests contains {
...
+NotGlobalAdmins contains User.DisplayName if {
+    some User in input.privileged_users
+    not "Global Administrator" in User.roles
+}
+
+default GetScoreDescription := "All privileged users are Global Admin"
+GetScoreDescription := concat("", ["Least Privilege Score = ", format_int(x,10), "%"]) if {
+    count(NotGlobalAdmins) > 0
+    x := count(GlobalAdmins)/count(NotGlobalAdmins)*100
+}
```
Comment (adhilto) [inline PowerShell/ScubaGear/Rego/AADConfig.rego:693]: One question about this. In the AAD baseline, the implementation tells you how to calculate the secure score, but the instructions over there produces a ratio, not a percentage, as in:
- The result isn't presented with a percent sign over there
- They don't multiple multiply the result by 100 over there
- They tell you that a passing score is 1 or less. Given that we're multiplying by 100 over here, you'd need to have at least 100 non-global admins for every global admin to achieve that. Probably not what we're going for :)

I don't think either formula is necessarily wrong, but the baseline implementation formula and Rego implementation formulas should match. I do have a slight preference for the ratio version over the percent version, as I think that's a bit more understandable.

Comment (adhilto) [reply PowerShell/ScubaGear/Rego/AADConfig.rego:693]: Optional: what if we also bake some guidance of how to interpret the score into the details column? One option: "Requirement failed: Least Privilege Score: 1.5 (should be 1 or less)"

**Review by adhilto — CHANGES_REQUESTED** [review]
Thanks for making the style changes I suggested. As I was testing just now though, I noticed a significant difference between how the secure score is calculated here and how it is described in the implementation section of the baseline, which needs to be addressed.

File: PowerShell/ScubaGear/Rego/AADConfig.rego (hunk 8 above)
Comment (ssatyapal123) [reply PowerShell/ScubaGear/Rego/AADConfig.rego]: Resolved (after great collaboration on slack) all except the final optional recommendation.

Comment (ssatyapal123) [reply PowerShell/ScubaGear/Rego/AADConfig.rego]: Added the extra text per suggestion.

**Review by adhilto — APPROVED** [review]

**Comment by tkol2022** [comment]
### Test Results

- Tested against G5, G3 and GCC high tenants to ensure that the score computes without errors - works fine
- Tested compliant and non-compliant scenarios against the GCC high tenant - results below - works as expected


#### Tested compliant scenario where there are less Global Admins than non Global Admins - produces a Pass as expected and calculation value is correct based on the numbers in the tenant
image

#### Tested case where 7.1 fails because there are too many Global Admins so 7.2 should automatically fail as well - produces a Fail as expected
image

#### Tested case where the number of Global Admins is equivalent to the number of non Global Admins - produces a Pass as expected because the calculation = 1
image


#### Tested non-compliant scenario where there are more Global Admins than non Global Admins - produces a policy Fail as expected
image

**Comment by tkol2022** [comment]
### Feedback on the Unit Tests

- Missing a test scenario for the case when the # of Global Admins is greater than 8. This will trigger a Fail of policy 7.1 which should automatically fail policy 7.2 since it is dependent on it.

- For test_SecureScore_Correct_V1 I'd like to see some variability in the usage of privileged roles referenced. Right now the test only uses Privileged Role Administrator for the non-global admin roles. Mix it up and usage a bunch of the other roles. The same applies to the other test scenarios for policy 7.2.

image

**Comment by tkol2022** [comment]
### Missing functional test

- Need functional test for policy 7.2

image

**Review by tkol2022 — CHANGES_REQUESTED** [review]
See my comments for requested changes and test results.

File: Testing/Functional/Products/TestPlans/aad.testplan.yaml [hunk 9]
```diff
@@ -771,13 +771,129 @@ TestPlan:
...
+          - Command: UpdateProviderExport
+            Splat:
+              Updates:
+                privileged_users:
+                  16b4d5c2-71c9-4644-8728-74e3a8324d81:
+                    DisplayName: Test User1
+                    roles:
+                      - Global Administrator
+                    OnPremisesImmutableId: null
+                    26b4d5c2-71c9-4644-8728-74e3a8324d82:
```
Comment (Sloane4) [inline Testing/Functional/Products/TestPlans/aad.testplan.yaml:844]:
```suggestion
                  26b4d5c2-71c9-4644-8728-74e3a8324d82:
```

File: Testing/Functional/Products/TestPlans/aad.testplan.yaml [hunk 10]
```diff
@@ -771,13 +771,129 @@ TestPlan:
...
+                    DisplayName: Test User2
+                    roles:
+                      - Global Administrator
+                      - Exchange Administrator
+                    OnPremisesImmutableId: null
+                    36b4d5c2-71c9-4644-8728-74e3a8324d83:
```
Comment (Sloane4) [inline Testing/Functional/Products/TestPlans/aad.testplan.yaml:850]:
```suggestion
                  36b4d5c2-71c9-4644-8728-74e3a8324d83:
```

**Review by Sloane4 — CHANGES_REQUESTED** [review]
Added corrections for what is causing your linter error

**Review by Sloane4 — APPROVED** [review]

**Review by tkol2022 — CHANGES_REQUESTED** [review]
A functional test scenario is missing: Compliant - Less number of Global Admins and other privileged users

**Review by tkol2022 — APPROVED** [review]
I made a slight update to the newly added test scenario name. Looks good to go.

**Comment by ssatyapal123** [comment]
@nanda-katikaneni this PR is ready to merge.
